// @benjeau/react-native-draw-extras/lib/commonjs/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SLIDERS_HEIGHT = exports.DEFAULT_TOOL = exports.DEFAULT_THICKNESS_STEP = exports.DEFAULT_THICKNESS_MIN = exports.DEFAULT_THICKNESS_MAX = exports.DEFAULT_THICKNESS = exports.DEFAULT_SLIDER_COLOR = exports.DEFAULT_OTHER_BUTTONS_COLOR = exports.DEFAULT_OPACITY_STEP = exports.DEFAULT_OPACITY = exports.DEFAULT_DELETE_BUTTON_COLOR = exports.DEFAULT_COLORS = exports.DEFAULT_BRUSH_PREVIEW = void 0;

var _types = require("./types");

const grayscale = [['#010101', '#151515', '#2A2A2A', '#3E3E3E', '#535353', '#666666', '#7A7A7A', '#8F8F8F', '#A3A3A3', '#B8B8B8', '#CCCCCC', '#FFFFFF']];
const colors = [['#457429', '#696F1C', '#6B6414', '#8C6115', '#8D5014', '#89280C', '#81090A', '#95133F', '#8204A6', '#3704AB', '#003389', '#0086AB'], ['#75BD4F', '#B5C835', '#C1BB2C', '#FFC934', '#FFAA2E', '#FF6924', '#F42C1B', '#DB4575', '#B644D0', '#7A4DD9', '#0063FA', '#00C7FC'], ['#A9D78A', '#E7F169', '#FFF755', '#FFD367', '#FFBE61', '#FF8F5B', '#FF6252', '#EC86A8', '#CE82E8', '#AB88F2', '#4897FA', '#5ADDFC'], ['#D6EFC3', '#F8FAB0', '#FFFA9E', '#FFE7AD', '#FFD9AC', '#FFC0A9', '#FFA9A7', '#FDC2D7', '#EDC1F9', '#D1C4F8', '#A2C4FB', '#AFEBFE']];
const DEFAULT_COLORS = [colors, grayscale];
exports.DEFAULT_COLORS = DEFAULT_COLORS;
const DEFAULT_THICKNESS = 3;
exports.DEFAULT_THICKNESS = DEFAULT_THICKNESS;
const DEFAULT_OPACITY = 1;
exports.DEFAULT_OPACITY = DEFAULT_OPACITY;
const DEFAULT_TOOL = _types.DrawingTool.Brush;
exports.DEFAULT_TOOL = DEFAULT_TOOL;
const DEFAULT_BRUSH_PREVIEW = 'stroke';
exports.DEFAULT_BRUSH_PREVIEW = DEFAULT_BRUSH_PREVIEW;
const DEFAULT_OPACITY_STEP = 0.1;
exports.DEFAULT_OPACITY_STEP = DEFAULT_OPACITY_STEP;
const DEFAULT_THICKNESS_MIN = 5;
exports.DEFAULT_THICKNESS_MIN = DEFAULT_THICKNESS_MIN;
const DEFAULT_THICKNESS_MAX = 35;
exports.DEFAULT_THICKNESS_MAX = DEFAULT_THICKNESS_MAX;
const DEFAULT_THICKNESS_STEP = 1;
exports.DEFAULT_THICKNESS_STEP = DEFAULT_THICKNESS_STEP;
const DEFAULT_SLIDER_COLOR = '#000';
exports.DEFAULT_SLIDER_COLOR = DEFAULT_SLIDER_COLOR;
const DEFAULT_DELETE_BUTTON_COLOR = '#81090A';
exports.DEFAULT_DELETE_BUTTON_COLOR = DEFAULT_DELETE_BUTTON_COLOR;
const DEFAULT_OTHER_BUTTONS_COLOR = '#DDD';
exports.DEFAULT_OTHER_BUTTONS_COLOR = DEFAULT_OTHER_BUTTONS_COLOR;
const SLIDERS_HEIGHT = 80;
exports.SLIDERS_HEIGHT = SLIDERS_HEIGHT;
